### PR TITLE
rustup to rustc 1.17.0-nightly (60a0edc6c 2017-02-26)

### PR DESCRIPTION
**Pull request by oli-obk**
We can now turn non-capturing closures into function pointers (and then back into `Fn` trait objects)

Diff best viewed [without whitespace changes](https://github.com/solson/miri/pull/147/files?w=1)

### CODE REVIEW

File: src/memory.rs [hunk 1]
```diff
@@ -238,6 +234,23 @@ impl<'a, 'tcx> Memory<'a, 'tcx> {
...
+            sig: fn_ty.sig,
+        });
+        self.create_fn_alloc(Function::NonCaptureClosureAsFnPtr(FunctionDefinition {
+            def_id,
+            substs: substs.substs,
+            abi: Abi::Rust, // adjust abi
+            // FIXME: why doesn't this compile?
+            //sig: tcx.erase_late_bound_regions(&fn_ty.sig),
+            sig: fn_ty.sig.skip_binder(),
+        }))
```
Comment (eddyb) [inline src/memory.rs]: I'm not sure this is the way to go. Did you see how this was implemented in `librustc_trans`?
That is, it relies on an existing shim- oh, drat, you have that extra argument to take care of, nevermind me!

**Comment by oli-obk** [comment]
Added rustup to @eddyb's [PR of ~~doom~~ awesomeness](https://github.com/rust-lang/rust/pull/40008)

File: src/memory.rs [hunk 2]
```diff
@@ -623,12 +596,7 @@ impl<'a, 'tcx> Memory<'a, 'tcx> {
 
 fn dump_fn_def<'tcx>(fn_def: FunctionDefinition<'tcx>) -> String {
     let name = ty::tls::with(|tcx| tcx.item_path_str(fn_def.def_id));
-    let abi = if fn_def.abi == Abi::Rust {
-        format!("")
-    } else {
-        format!("extern {} ", fn_def.abi)
-    };
-    format!("function pointer: {}: {}{}", name, abi, fn_def.sig)
+    format!("function pointer: {}: {}", name, fn_def.sig.skip_binder())
```
Comment (eddyb) [inline src/memory.rs]: I don't think skipping the binder is needed here at all!

Comment (oli-obk) [inline src/memory.rs]: yes it is :) There's no display impl for Binder

File: src/eval_context.rs [hunk 3]
```diff
@@ -196,7 +194,7 @@ impl<'a, 'tcx> EvalContext<'a, 'tcx> {
 
             Struct(_)    => unimplemented!(),
             Tuple(_)     => unimplemented!(),
-            Function(_)  => unimplemented!(),
+            Function(_, _)  => unimplemented!(),
```
Comment (eddyb) [inline src/eval_context.rs]: Tempted to make this the way to refer to a `TyFnDef` (now that it has `Substs`), what do you think?

Comment (oli-obk) [inline src/eval_context.rs]: it's unreachable ^^ At least in the run-pass tests that we run. We only use this function in `mir::Literal::Value`, and I think that can only contain primitives

Comment (eddyb) [inline src/eval_context.rs]: Oh, I know, it's because we use `Item` right now, and work around the awkwardness of it everywhere.

Comment (oli-obk) [inline src/eval_context.rs]: Oh... you mean changing it that `Literal::Item` stops referring to functions. Sounds good.

File: src/memory.rs (hunk 2 above)
Comment (eddyb) [inline src/memory.rs]: Oh, right, it's `Debug` only - should still do the right thing though.

File: src/lvalue.rs [hunk 4]
```diff
@@ -137,9 +137,9 @@ impl<'a, 'tcx> EvalContext<'a, 'tcx> {
             Local(mir::RETURN_POINTER) => self.frame().return_lvalue,
             Local(local) => Lvalue::Local { frame: self.stack.len() - 1, local, field: None },
 
-            Static(def_id) => {
+            Static(ref statik) => {
```
Comment (oli-obk) [inline src/lvalue.rs]: sgtm, I was just following the style seen in rustc (`krate`)

File: src/terminator/mod.rs [hunk 5]
```diff
@@ -61,35 +61,54 @@ impl<'a, 'tcx> EvalContext<'a, 'tcx> {
...
-                        // FIXME: also check the size of the arguments' type and the return type
-                        // Didn't get it to work, since that triggers an assertion in rustc which
-                        // checks whether the type has escaping regions
-                        if abi != bare_fn_ty.abi ||
-                           sig.variadic != bare_sig.variadic ||
-                           sig.inputs().len() != bare_sig.inputs().len() {
-                            return Err(EvalError::FunctionPointerTyMismatch(abi, sig, bare_fn_ty));
+                        match fn_def {
+                            Function::Concrete(fn_def) => {
+                                // transmuting function pointers in miri is fine as long as the number of
```
Comment (oli-obk) [inline src/terminator/mod.rs]: yea... I tried to do that check, but since we also don't care about lifetime changes I tried to erase the lifetimes, which kept causing rustc to panic in weird corner cases.

I'll change the FIXME text

Comment (oli-obk) [inline src/terminator/mod.rs]: oh... I remember sth else. Rustc tests whether we can transmute `fn foo(*const i32)` to `fn foo(Option<&i32>)` (or maybe the other direction).

Comment (solson) [inline src/terminator/mod.rs]: I'd prefer to be more conservative than rustc about function pointer casts until a more formal definition comes along, unless it becomes essential for evaluating std code.

Comment (oli-obk) [inline src/terminator/mod.rs]: [reran on rust test suite:](https://github.com/solson/miri/issues/55)

```
   1 tried to call a function with sig fn(&[u8]) -> std::option::Option<u8> through a function pointer of type fn(&[u8]) -> std::option::Option<u8>
   1 tried to call a function with sig fn(&A<'_>) through a function pointer of type fn(&A)
   1 tried to call a function with sig extern "C" fn(usize) -> Foo through a function pointer of type extern "C" fn(usize) -> u32
   1 tried to call a function with sig extern "C" fn(&isize) -> std::option::Option<&isize> through a function pointer of type extern "C" fn(&isize) -> &isize
```

Comment (solson) [inline src/terminator/mod.rs]: 1. `fn(&[u8]) -> std::option::Option<u8>` => `fn(&[u8]) -> std::option::Option<u8>`
2. `fn(&A<'_>)` => `fn(&A)`
3. `extern "C" fn(usize) -> Foo` => `extern "C" fn(usize) -> u32`
4. `extern "C" fn(&isize) -> std::option::Option<&isize>` => `extern "C" fn(&isize) -> &isize`

Of these, I think we should make sure 1 and 2 work (ignore lifetimes). I don't care about supporting 3 and 4 for now.

File: src/eval_context.rs [hunk 6]
```diff
@@ -225,6 +225,13 @@ impl<'a, 'tcx> EvalContext<'a, 'tcx> {
         self.tcx.normalize_associated_type(&substituted)
     }
 
+    pub fn erase_lifetimes<T>(self, value: &Binder<T>) -> T
+        where T : TypeFoldable<'tcx>
+    {
+        let value = self.tcx.erase_late_bound_regions(value)
```
Comment (solson) [inline src/eval_context.rs]: Missing semicolon here.

**Comment by oli-obk** [comment]
fixed
